clang/lib/Lex/Preprocessor.cpp

```diff
@@ -1329,6 +1341,100 @@ bool Preprocessor::LexAfterModuleImport(Token &Result) {
+///             pp-module-name-qualifier identifier .
+/// No identifier in the pp-module-name or pp-module-partition shall currently
+/// be defined as an object-like macro.
+///
+/// [cpp.module]/p3:
+/// Any preprocessing tokens after the module preprocessing token in the module
+/// directive are processed just as in normal text.
+bool Preprocessor::LexAfterModuleDecl(Token &Result) {
+  // Figure out what kind of lexer we actually have.
+  recomputeCurLexerKind();
+  LexUnexpandedToken(Result);
+
+  auto EnterTokens = [this](ArrayRef<Token> Toks, bool DisableMacroExpansion) {
+    auto ToksCopy = std::make_unique<Token[]>(Toks.size());
+    std::copy(Toks.begin(), Toks.end(), ToksCopy.get());
+    EnterTokenStream(std::move(ToksCopy), Toks.size(), DisableMacroExpansion,
+                     /*IsReinject=*/false);
+  };
+
+  // If we don't expect an identifier but got an identifier, it's not a part of
+  // module name.
+  if (!ModuleDeclExpectsIdentifier && Result.is(tok::identifier)) {
+    EnterTokens(Result, /*DisableMacroExpansion=*/false);
+    return false;
+  }
+
+  // The token sequence
+  //
+  // export[opt] module identifier (. identifier)*
+  //
+  // indicates a module directive. We already saw the 'module'
+  // contextual keyword, so now we're looking for the identifiers.
+  if (ModuleDeclExpectsIdentifier && Result.is(tok::identifier)) {
+    auto *MI = getMacroInfo(Result.getIdentifierInfo());
+    if (MI && MI->isObjectLike()) {
+      Diag(Result, diag::err_module_decl_cannot_be_macros)
+          << Result.getLocation() << ModuleDeclLexingPartitionName
+          << Result.getIdentifierInfo();
+    }
+    ModuleDeclExpectsIdentifier = false;
+    CurLexerCallback = CLK_LexAfterModuleDecl;
+    return true;
+  }
+
+  // If we're expecting a '.', a ':' or a ';', and we got a '.', then wait until
+  // we see the next identifier.
+  if (!ModuleDeclExpectsIdentifier && Result.isOneOf(tok::period, tok::colon)) {
+    ModuleDeclExpectsIdentifier = true;
+    ModuleDeclLexingPartitionName = Result.is(tok::colon);
+    CurLexerCallback = CLK_LexAfterModuleDecl;
+    return true;
+  }
+
+  // [cpp.module]/p2: where the pp-tokens (if any) shall not begin with a (
+  // preprocessing token [...]
+  if (!ModuleDeclExpectsIdentifier && Result.is(tok::l_paren)) {
+    ModuleDeclExpectsIdentifier = false;
+    Diag(Result, diag::err_unxepected_paren_in_module_decl)
+        << ModuleDeclLexingPartitionName;
+    Token Tok;
+    // We already have a '('.
+    unsigned NumParens = 1;
+    while (true) {
+      LexUnexpandedToken(Tok);
+      if (Tok.isOneOf(tok::eod, tok::eof, tok::semi, tok::period, tok::colon)) {
+        EnterTokens(Tok, /*DisableMacroExpansion=*/true);
+        break;
+      }
+      if (Tok.is(tok::l_paren))
+        NumParens++;
+      else if (Tok.is(tok::r_paren) && --NumParens == 0)
+        break;
+    }
+    CurLexerCallback = CLK_LexAfterModuleDecl;
+    return false;
+  }
+
+  return true;
+}
+
```

Comment:
What is the motivation to put the logic in the preprocessor ? It does adds quite a bit of complexity to the state machine.
But I struggle to make up my mind on what the better approach is.

I kept coming back to the idea of using an annotation

- Introduce a new `cxx_module_name` token annotation in `TokenKind.defs`
- Parse the full module name in the preprocessor
- Store it in  a `Preprocessor` or `Lexer` member (we need 2 `SmallVectorImpl<std::pair<IdentifierInfo *, SourceLocation>>`, one for the partition, one for the name)
- The value of that annotation would be a pointer to that pair of vector
- `ParseModuleName` would handle `cxx_module_name` by moving the content of the vector.
(It w

- That way `LexAfterModuleDecl` can just produce a single token with a fully parsed module name and we avoid having the logic spread in multiple places with complex state machines.


---

Alternatively, and perhaps much simpler,  lex the whole module name + partition at once here do all the checks and diagnostics and reinject all of these tokens all at once.

... actually, that the second solution should work nicely, right?


WDYT?

Reply:
> What is the motivation to put the logic in the preprocessor ? It does adds quite a bit of complexity to the state machine.

There are two reasons. The 1st is to be consistent with the way that `import` was processed. The 2nd reason is that some steps in Parser will look ahead of the token before ParseModuleDecl. In ParseModuleName, causing the first identifier was macro expanded token.  So once we saw a `module` keyword, The next token and the token conforming to the form `identifier(.identifier)*` must be processed in `Preprocessor::LexAfterModuleDecl`.

Introduce a new `cxx_module_name` also looks good to me, it's looks like the approach processing `#pragma`.

Reply:
> Introduce a new cxx_module_name also looks good to me, it's looks like the approach processing #pragma.

Let's try to do that then

Reply:
Sorry for the very late reply! I’ve tried to implement this approach. But I’m fall into trouble. 

```
#ifndef VERSION_H
#define VERSION_H

#define VERSION libv5
#define A a
#define B b
#define C c
#define FUNC_LIKE(X) function_like_##X
#define ATTR [[]]
#define SEMICOLON ;

#endif


export module a.FUNC_LIKE:c ATTRS; // OK, FUNC_LIKE would not be treated as a macro name.
```

*The 1st approach*:

We try to consume all module name tokens and return a token::cxx_module_name, which includes {‘a’, ‘.’, ‘FUNC_LIKE’, ‘.’, ‘:’, ‘c’}, and we also will consumed and stop at ‘ATTRS’, but it’s not a part of module name, and need to be put back to token stream. Current a Lex action cannot both return a token and cache tokens (doing so would corrupt the token cache if the call to Lex comes from CachingLex / PeekAhead).

*The 2nd approach*:

We try to consume all module name tokens and got {‘a’, ‘.’, ‘FUNC_LIKE’, ‘.’, ‘:’, ‘c’} we need to put this token array back to token stream with macro expansion disabled. Also, we have consumed and stop at ‘ATTRS’, but it’s not a part of module name, and need to be put back to token stream(enable macro expansion). The two EnterTokenStreans are conflict.

I also tried to eliminate the complex state machine in LexAfterModuleImport.

Reply:
friendly ping~

Reply:
Friendly ping~

Reply:
I am not sure I under why we would actually consume `ATTRS` in either case, but I think I need to think about it more / play with the PR.

@AaronBallman Any opinion?

Reply:
> We try to consume all module name tokens and return a token::cxx_module_name, which includes {‘a’, ‘.’, ‘FUNC_LIKE’, ‘.’, ‘:’, ‘c’}, and we also will consumed and stop at ‘ATTRS’, but it’s not a part of module name, and need to be put back to token stream. Current a Lex action cannot both return a token and cache tokens (doing so would corrupt the token cache if the call to Lex comes from CachingLex / PeekAhead).

I've been thinking about that.
I think _both_ the annotation and the peeked token need to be re-entered
and then you can call PP.Lex(Result) again and return.

Is that something you would be willing to explore?

Reply:
> I am not sure I under why we would actually consume ATTRS in either case, but I think I need to think about it more / play with the PR.

I'd like to understand this better, too. I would have (naively and without checking in a debugger) that we would have peeked at `ATTRS` rather than consumed it.

> @AaronBallman Any opinion?

I think your suggestion makes sense to me; that's a good use of annotation tokens to squirrel information from the preprocessor into the parser, and keeping the logic consolidated as much as possible is preferred.